contrib/src/msgpack/mod.rs

```diff
@@ -0,0 +1,118 @@
+extern crate rmp_serde;
+
+use std::ops::{Deref, DerefMut};
+use std::io::Read;
+
+use rocket::outcome::Outcome;
+use rocket::request::Request;
+use rocket::data::{self, Data, FromData};
+use rocket::response::{self, Responder, content};
+use rocket::http::Status;
+
+use serde::{Serialize, Deserialize};
+
+pub use self::rmp_serde::decode::Error as MsgPackError;
+
+/// The `MsgPack` type: implements `FromData` and `Responder`, allowing you to easily
+/// consume and respond with MessagePack.
+///
+/// If you're receiving MessagePack data, simply add a `data` parameter to your route
+/// arguments and ensure the type of the parameter is a `MsgPack<T>`, where `T` is
+/// some type you'd like to parse from MessagePack. `T` must implement `Deserialize`
+/// from [Serde](https://github.com/serde-rs/serde). The data is parsed from the
+/// HTTP request body.
+///
+/// ```rust,ignore
+/// #[post("/users/", format = "application/msgpack", data = "<user>")]
+/// fn new_user(user: MsgPack<User>) {
+///     ...
+/// }
+/// ```
+/// You don't _need_ to use `format = "application/msgpack"`, but it _may_ be what
+/// you want. Using `format = application/msgpack` means that any request that
+/// doesn't specify "application/msgpack" as its first `Content-Type:` header
+/// parameter will not be routed to this handler.
+///
+/// If you're responding with MessagePack data, return a `MsgPack<T>` type, where `T`
+/// implements `Serialize` from [Serde](https://github.com/serde-rs/serde). The
+/// content type of the response is set to `application/msgpack` automatically.
+///
+/// ```rust,ignore
+/// #[get("/users/<id>")]
+/// fn user(id: usize) -> MsgPack<User> {
+///     let user_from_id = User::from(id);
+///     ...
+///     MsgPack(user_from_id)
+/// }
+/// ```
+///
+#[derive(Debug)]
+pub struct MsgPack<T>(pub T);
+
+impl<T> MsgPack<T> {
+    /// Consumes the `MsgPack` wrapper and returns the wrapped item.
+    ///
+    /// # Example
+    /// ```rust
+    /// # use rocket_contrib::MsgPack;
+    /// let string = "Hello".to_string();
+    /// let my_msgpack = MsgPack(string);
+    /// assert_eq!(my_msgpack.into_inner(), "Hello".to_string());
+    /// ```
+    pub fn into_inner(self) -> T {
+        self.0
+    }
+}
+
+impl<T: Deserialize> FromData for MsgPack<T> {
+    type Error = MsgPackError;
+
+    fn from_data(request: &Request, data: Data) -> data::Outcome<Self, Self::Error> {
+        if !request.content_type().map_or(false, |ct| ct.is_msgpack()) {
+            error_!("Content-Type is not MessagePack.");
+            return Outcome::Forward(data);
+        }
+
+        let mut buf = Vec::new();
+        if let Err(e) = data.open().read_to_end(&mut buf) {
```

Comment:
This is bad as it opens the application up to DOS attack via an out-of-memory problem. The client can just stream infinite data until the machine runs out of memory. This needs to be limited.